File: rust/src/spider_browser.rs
//! SpiderBrowser — main entry point for spider-browser.

use crate::errors::{Result, SpiderError};
use crate::events::SpiderEventEmitter;
use crate::page::SpiderPage;
use crate::protocol::protocol_adapter::{ProtocolAdapter, ProtocolAdapterOptions};
use crate::protocol::transport::{Transport, TransportOptions};
use arc_swap::ArcSwap;
use std::sync::Arc;
use tokio::sync::mpsc;
use tracing::info;

#[cfg(feature = "ai")]
use crate::ai::llm_provider::{create_provider, LLMConfig, LLMProvider};

/// Options for creating a SpiderBrowser instance.
#[derive(Clone, Debug)]
pub struct SpiderBrowserOptions {
    /// Spider API key (required).
    pub api_key: String,
    /// WebSocket server URL (default: wss://browser.spider.cloud).
    pub server_url: Option<String>,
    /// Browser to use (default: "auto").
    pub browser: Option<String>,
    /// Target URL hint for server browser+proxy selection.
    pub url: Option<String>,
    /// Captcha handling: "off", "detect", or "solve" (default: "solve").
    pub captcha: Option<String>,
    /// Enable smart retry with browser switching (default: true).
    pub smart_retry: Option<bool>,
    /// Max retry attempts across all browsers (default: 12).
    pub max_retries: Option<u32>,
    /// Stealth level (1-3). 0 = auto-escalate on failure.
    pub stealth: Option<u32>,
    /// Maximum stealth level to auto-escalate to (1-3, default: 3).
    pub max_stealth_levels: Option<u32>,
    /// WebSocket connect timeout in ms (default: 30000).
    pub connect_timeout_ms: Option<u64>,
    /// CDP/BiDi command timeout in ms (default: 30000).
    pub command_timeout_ms: Option<u64>,
    /// Timeout for retry attempts (default: 15000).
    pub retry_timeout_ms: Option<u64>,
    /// Mark this session as a hedge (parallel attempt).
    pub hedge: Option<bool>,
    /// Enable screencast recording (default: false).
    pub record: Option<bool>,
    /// Browser mode: "scraping" or "cua".
    pub mode: Option<String>,
    /// LLM configuration for AI methods.
    #[cfg(feature = "ai")]
    pub llm: Option<LLMConfig>,
}

impl SpiderBrowserOptions {
    pub fn new(api_key: impl Into<String>) -> Self {
        Self {
            api_key: api_key.into(),
            server_url: None,
            browser: None,
            url: None,
            captcha: None,
            smart_retry: None,
            max_retries: None,
            stealth: None,
            max_stealth_levels: None,
            connect_timeout_ms: None,
            command_timeout_ms: None,
            retry_timeout_ms: None,
            hedge: None,
            record: None,
            mode: None,
            #[cfg(feature = "ai")]
            llm: None,
        }
    }
}

struct ResolvedOptions {
    api_key: String,
    server_url: String,
    browser: String,
    url: Option<String>,
    captcha: String,
    smart_retry: bool,
    max_retries: u32,
    stealth: u32,
    max_stealth_levels: u32,
    connect_timeout_ms: u64,
    command_timeout_ms: u64,
    retry_timeout_ms: u64,
    hedge: Option<bool>,
    record: Option<bool>,
    mode: Option<String>,
}

/// SpiderBrowser — connects to Spider's pre-warmed browser fleet.
///
/// Provides deterministic page control (via SpiderPage) and
/// AI-powered automation (act, observe, extract, agent).
pub struct SpiderBrowser {
    opts: ResolvedOptions,
    transport: Option<Arc<Transport>>,
    adapter: Option<Arc<ProtocolAdapter>>,
    page: Option<Arc<SpiderPage>>,
    emitter: SpiderEventEmitter,
    current_url: ArcSwap<Option<String>>,
    #[cfg(feature = "ai")]
    llm_provider: Option<Box<dyn LLMProvider>>,
    msg_send_tx: Option<mpsc::UnboundedSender<String>>,
}

impl SpiderBrowser {
    pub fn new(options: SpiderBrowserOptions) -> Self {
        let resolved = ResolvedOptions {
            api_key: options.api_key.clone(),
            server_url: options.server_url.unwrap_or_else(|| "wss://browser.spider.cloud".into()),
            browser: options.browser.unwrap_or_else(|| "auto".into()),
            url: options.url.clone(),
            captcha: options.captcha.unwrap_or_else(|| "solve".into()),
            smart_retry: options.smart_retry.unwrap_or(true),
            max_retries: options.max_retries.unwrap_or(12),
            stealth: options.stealth.unwrap_or(0),
            max_stealth_levels: options.max_stealth_levels.unwrap_or(3),
            connect_timeout_ms: options.connect_timeout_ms.unwrap_or(30_000),
            command_timeout_ms: options.command_timeout_ms.unwrap_or(30_000),
            retry_timeout_ms: options.retry_timeout_ms.unwrap_or(15_000),
            hedge: options.hedge,
            record: options.record,
            mode: options.mode.clone(),
        };

        #[cfg(feature = "ai")]
        let llm_provider: Option<Box<dyn LLMProvider>> =
            options.llm.map(|config| create_provider(config));

        Self {
            opts: resolved,
            transport: None,
            adapter: None,
            page: None,
            emitter: SpiderEventEmitter::new(),
            current_url: ArcSwap::from_pointee(options.url),
            #[cfg(feature = "ai")]
            llm_provider,
            msg_send_tx: None,
        }
    }

    /// The active page instance for deterministic browser control.
    pub fn page(&self) -> &SpiderPage {
        self.page
            .as_ref()
            .expect("SpiderBrowser not initialized. Call init() first.")
    }

    /// Current browser type.
    pub fn browser(&self) -> String {
        self.transport
            .as_ref()
            .map(|t| t.browser())
            .unwrap_or_else(|| self.opts.browser.clone())
    }

    /// Whether the WebSocket is connected.
    pub fn connected(&self) -> bool {
        self.transport
            .as_ref()
            .map(|t| t.is_connected())
            .unwrap_or(false)
    }

    /// Active stealth level.
    pub fn stealth_level(&self) -> u32 {
        self.transport
            .as_ref()
            .map(|t| t.get_stealth_level())
            .unwrap_or(self.opts.stealth)
    }

    /// Credits remaining from last upgrade response.
    pub fn credits(&self) -> Option<f64> {
        self.transport.as_ref().and_then(|t| t.upgrade_credits())
    }

    /// Credits consumed during this session.
    pub fn session_credits_used(&self) -> Option<f64> {
        self.transport
            .as_ref()
            .and_then(|t| t.session_credits_used())
    }

    /// Subscribe to events.
    pub fn on(&self, event: &str, handler: crate::events::EventHandler) {
        self.emitter.on(event, handler);
    }

    /// Connect to the browser fleet and initialize the protocol.
    pub async fn init(&mut self) -> Result<()> {
        let transport_opts = TransportOptions {
            api_key: self.opts.api_key.clone(),
            server_url: self.opts.server_url.clone(),
            browser: self.opts.browser.clone(),
            url: self.opts.url.clone(),
            captcha: Some(self.opts.captcha.clone()),
            stealth_level: self.opts.stealth,
            connect_timeout_ms: self.opts.connect_timeout_ms,
            command_timeout_ms: self.opts.command_timeout_ms,
            hedge: self.opts.hedge.unwrap_or(false),
            record: self.opts.record.unwrap_or(false),
            mode: self.opts.mode.clone(),
        };

        let transport = Transport::new(transport_opts, self.emitter.clone());
        transport.connect(3).await?;

        let active_browser = transport.browser();

        // Take the message receiver and set up routing
        let mut msg_rx = transport
            .take_message_rx()
            .await
            .ok_or_else(|| SpiderError::Protocol("Message receiver already taken".into()))?;

        let adapter_opts = if self.opts.command_timeout_ms != 30_000 {
            Some(ProtocolAdapterOptions {
                command_timeout_ms: Some(self.opts.command_timeout_ms),
            })
        } else {
            None
        };

        // Create a relay channel that forwards outgoing messages to the transport's WS
        let (proto_tx, mut proto_rx) = mpsc::unbounded_channel::<String>();
        let transport_for_relay = Arc::clone(&transport);
        tokio::spawn(async move {
            while let Some(data) = proto_rx.recv().await {
                let _ = transport_for_relay.send(data);
            }
        });

        let mut adapter = ProtocolAdapter::new(
            proto_tx.clone(),
            self.emitter.clone(),
            &active_browser,
            adapter_opts,
        );
        adapter.init().await?;

        // Wrap adapter in Arc for sharing between page and AI methods
        let adapter = Arc::new(adapter);
        let page = SpiderPage::from_arc(Arc::clone(&adapter));
        let page = Arc::new(page);

        // Spawn message routing task: transport -> adapter.route_message()
        let page_for_routing = Arc::clone(&page);
        tokio::spawn(async move {
            while let Some(data) = msg_rx.recv().await {
                page_for_routing.route_message(&data);
            }
        });

        self.transport = Some(transport);
        self.adapter = Some(adapter);
        self.page = Some(page);
        self.msg_send_tx = Some(proto_tx);

        info!("SpiderBrowser initialized (browser={})", active_browser);
        Ok(())
    }

    /// Navigate to a URL with smart retry.
    pub async fn goto(&self, url: &str) -> Result<()> {
        self.current_url.store(Arc::new(Some(url.to_string())));
        self.page().goto(url).await
    }

    /// Close the connection and clean up resources.
    pub fn close(&mut self) {
        if let Some(ref page) = self.page {
            page.destroy();
        }
        if let Some(ref transport) = self.transport {
            transport.close();
        }
        self.emitter.remove_all_listeners();
        self.page = None;
        self.adapter = None;
        self.transport = None;
        info!("SpiderBrowser closed");
    }

    // ------------------------------------------------------------------
    // AI Methods (require "ai" feature + LLM config)
    // ------------------------------------------------------------------

    /// Execute a single action from natural language.
    #[cfg(feature = "ai")]
    pub async fn act(&self, instruction: &str) -> Result<()> {
        let llm = self.require_llm()?;
        let adapter = self.require_adapter()?;
        crate::ai::act::act(adapter, llm.as_ref(), instruction).await
    }

    /// Discover interactive elements on the page.
    #[cfg(feature = "ai")]
    pub async fn observe(
        &self,
        instruction: Option<&str>,
    ) -> Result<Vec<crate::ai::observe::ObserveResult>> {
        let adapter = self.require_adapter()?;
        let llm_ref: Option<&dyn LLMProvider> = self.llm_provider.as_ref().map(|b| b.as_ref());
        crate::ai::observe::observe(adapter, instruction, llm_ref).await
    }

    /// Extract structured data from the page.
    #[cfg(feature = "ai")]
    pub async fn extract<T: serde::de::DeserializeOwned + Send>(
        &self,
        instruction: &str,
    ) -> Result<T> {
        let llm = self.require_llm()?;
        let adapter = self.require_adapter()?;
        crate::ai::extract::extract(adapter, llm.as_ref(), instruction, None).await
    }

    /// Create an autonomous agent.
    #[cfg(feature = "ai")]
    pub fn agent(
        &self,
        options: Option<crate::ai::agent::AgentOptions>,
    ) -> crate::ai::agent::Agent<'_> {
        let llm = self
            .llm_provider
            .as_ref()
            .expect("LLM not configured. Pass llm option for AI methods.");
        let adapter = self
            .adapter
            .as_ref()
            .expect("SpiderBrowser not initialized. Call init() first.");
        crate::ai::agent::Agent::new(adapter, llm.as_ref(), &self.emitter, options)
    }

    #[cfg(feature = "ai")]
    fn require_llm(&self) -> Result<&Box<dyn LLMProvider>> {
        self.llm_provider.as_ref().ok_or_else(|| {
            SpiderError::Llm(
                "LLM not configured. Pass llm option to SpiderBrowser for AI methods.".into(),
            )
        })
    }

    fn require_adapter(&self) -> Result<&ProtocolAdapter> {
        self.adapter
            .as_ref()
            .map(|a| a.as_ref())
            .ok_or_else(|| {
                SpiderError::Protocol(
                    "SpiderBrowser not initialized. Call init() first.".into(),
                )
            })
    }
}

File: python/spider_browser/spider_browser.py
"""SpiderBrowser — main entry point for spider-browser Python client."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, List, Literal, Optional, TypeVar

from .protocol.transport import Transport
from .protocol.protocol_adapter import ProtocolAdapter
from .events.emitter import SpiderEventEmitter
from .events.types import BrowserType
from .page import SpiderPage
from .retry.retry_engine import RetryEngine, RetryContext
from .ai.llm_provider import LLMConfig, LLMProvider, create_provider
from .ai.act import act
from .ai.observe import observe, ObserveResult
from .ai.extract import extract
from .ai.agent import Agent, AgentOptions, AgentResult
from .utils.logger import logger, set_level

T = TypeVar("T")


@dataclass
class SpiderBrowserOptions:
    """Configuration for SpiderBrowser."""

    api_key: str
    server_url: str = "wss://browser.spider.cloud"
    browser: BrowserType = "auto"
    url: Optional[str] = None
    captcha: Literal["off", "detect", "solve"] = "solve"
    smart_retry: bool = True
    max_retries: int = 12
    stealth: int = 0
    max_stealth_levels: int = 3
    log_level: Optional[str] = None
    connect_timeout_ms: int = 30_000
    command_timeout_ms: int = 30_000
    retry_timeout_ms: int = 15_000
    hedge: bool = False
    record: bool = False
    mode: Optional[str] = None  # 'scraping' or 'cua'
    country: Optional[str] = None
    proxy_url: Optional[str] = None
    llm: Optional[LLMConfig] = None


class SpiderBrowser:
    """
    Main entry point for spider-browser.

    Connects to Spider's pre-warmed browser fleet via WebSocket.
    Provides deterministic page control (via SpiderPage) and
    AI-powered automation (act, observe, extract, agent).

    Key features:
    - Pre-warmed browsers (no cold start)
    - Full stealth Chrome & Firefox
    - Smart retry with automatic browser switching
    - CDP (Chrome/Servo/LightPanda) + BiDi (Firefox) protocol support
    """

    def __init__(self, options: SpiderBrowserOptions) -> None:
        self._opts = options
        self._transport: Optional[Transport] = None
        self._adapter: Optional[ProtocolAdapter] = None
        self._retry_engine: Optional[RetryEngine] = None
        self._emitter = SpiderEventEmitter()
        self._page: Optional[SpiderPage] = None
        self._llm_provider: Optional[LLMProvider] = None
        self._current_url: Optional[str] = options.url

        if options.log_level:
            set_level(options.log_level)

        if options.llm:
            self._llm_provider = create_provider(options.llm)

    @property
    def page(self) -> SpiderPage:
        """The active page instance for deterministic browser control."""
        if not self._page:
            raise RuntimeError("SpiderBrowser not initialized. Call init() first.")
        return self._page

    @property
    def browser(self) -> BrowserType:
        """Current browser type."""
        if self._transport:
            return self._transport.browser
        return self._opts.browser

    @property
    def connected(self) -> bool:
        return self._transport.connected if self._transport else False

    @property
    def stealth_level(self) -> int:
        """Active stealth level (0=auto, 1-3=explicit tiers)."""
        if self._retry_engine:
            return self._retry_engine.stealth_level
        if self._transport:
            return self._transport.stealth_level
        return self._opts.stealth

    @property
    def credits(self) -> Optional[float]:
        """Credits remaining from last upgrade response."""
        return self._transport.upgrade_credits if self._transport else None

    @property
    def session_credits_used(self) -> Optional[float]:
        """Credits consumed during this session (from server Spider.metering event)."""
        return self._transport.session_credits_used if self._transport else None

    async def get_session_credits(self) -> float:
        """Request the exact session cost from the server.
        Unlike session_credits_used (which relies on async event delivery),
        this sends a Spider.getMetering command and waits for the response.
        Call this before close() for accurate per-session metering.
        """
        if not self._transport:
            return 0
        return await self._transport.request_metering()

    def on(self, event: str, handler: Any) -> "SpiderBrowser":
        """Subscribe to events."""
        self._emitter.on(event, handler)
        return self

    def off(self, event: str, handler: Any) -> "SpiderBrowser":
        """Unsubscribe from events."""
        self._emitter.off(event, handler)
        return self

    def once(self, event: str, handler: Any) -> "SpiderBrowser":
        """Subscribe to an event once."""
        self._emitter.once(event, handler)
        return self

    async def init(self) -> None:
        """Connect to the browser server WebSocket and initialize the protocol."""
        self._transport = Transport(
            api_key=self._opts.api_key,
            server_url=self._opts.server_url,
            browser=self._opts.browser,
            url=self._opts.url,
            captcha=self._opts.captcha,
            stealth=self._opts.stealth,
            record=self._opts.record,
            mode=self._opts.mode,
            hedge=self._opts.hedge,
            country=self._opts.country,
            proxy_url=self._opts.proxy_url,
            connect_timeout_ms=self._opts.connect_timeout_ms,
            command_timeout_ms=self._opts.command_timeout_ms,
            emitter=self._emitter,
        )
        await self._transport.connect()

        active_browser = self._transport.browser
        self._adapter = ProtocolAdapter(
            self._transport,
            self._emitter,
            active_browser,
            command_timeout_ms=self._opts.command_timeout_ms,
        )
        await self._adapter.init()

        self._page = SpiderPage(self._adapter)

        if self._opts.smart_retry:
            self._retry_engine = RetryEngine(
                max_retries=self._opts.max_retries,
                emitter=self._emitter,
                retry_timeout_ms=self._opts.retry_timeout_ms,
                command_timeout_ms=self._opts.command_timeout_ms,
                max_stealth_level=self._opts.max_stealth_levels,
                initial_stealth=self._opts.stealth,
            )

        self._current_url = self._opts.url
        logger.info(f"SpiderBrowser initialized (browser={active_browser})")

    async def with_retry(self, fn: Callable[[], Awaitable[T]]) -> T:
        """Execute with smart retry and browser switching."""
        if not self._retry_engine or not self._transport or not self._adapter:
            return await fn()

        ctx = RetryContext(
            transport=self._transport,
            adapter=self._adapter,
            current_url=self._current_url,
            on_adapter_changed=self._on_adapter_changed,
        )
        return await self._retry_engine.execute(fn, ctx)

    # -------------------------------------------------------------------
    # Navigation (with retry)
    # -------------------------------------------------------------------

    async def goto(self, url: str) -> None:
        """Navigate to a URL with smart retry.

        On ERR_ABORTED: closes the WebSocket, reconnects, and retries.
        On bot detection: switches to a different browser and retries.
        Also updates current_url for subsequent retries on other operations.
        """
        self._current_url = url
        await self.with_retry(lambda: self._page.goto(url))  # type: ignore

    # -------------------------------------------------------------------
    # AI Methods (require LLM config)
    # -------------------------------------------------------------------

    async def act(self, instruction: str) -> None:
        """Execute a single action from natural language."""
        self._require_llm()
        await self.with_retry(lambda: act(self._adapter, self._llm_provider, instruction))  # type: ignore

    async def observe(self, instruction: Optional[str] = None) -> List[ObserveResult]:
        """Discover interactive elements on the page."""
        return await self.with_retry(
            lambda: observe(self._adapter, instruction, self._llm_provider)  # type: ignore
        )

    async def extract(self, instruction: str, schema: Any = None) -> Any:
        """Extract structured data from the page."""
        self._require_llm()
        return await self.with_retry(
            lambda: extract(self._adapter, self._llm_provider, instruction, schema)  # type: ignore
        )

    def agent(self, options: Optional[AgentOptions] = None) -> Agent:
        """Create an autonomous agent for multi-step tasks."""
        self._require_llm()
        return Agent(self._adapter, self._llm_provider, self._emitter, options)  # type: ignore

    async def close(self) -> None:
        """Close the connection and clean up resources.

        Disconnects the WebSocket. The server automatically handles page/target
        cleanup on disconnect (disposes browser contexts, closes orphaned targets,
        deletes backend sessions). The browser process itself is NOT closed —
        it's pre-warmed and shared.
        """
        if self._adapter:
            self._adapter.destroy()
        if self._transport:
            await self._transport.close()
        self._emitter.remove_all_listeners()
        self._page = None
        self._adapter = None
        self._transport = None
        logger.info("SpiderBrowser closed")

    # -------------------------------------------------------------------
    # Context manager
    # -------------------------------------------------------------------

    async def __aenter__(self) -> "SpiderBrowser":
        await self.init()
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.close()

    # -------------------------------------------------------------------
    # Internal
    # -------------------------------------------------------------------

    def _on_adapter_changed(self, new_adapter: ProtocolAdapter) -> None:
        self._adapter = new_adapter
        if self._page:
            self._page._set_adapter(new_adapter)

    def _require_llm(self) -> None:
        if not self._llm_provider:
            raise RuntimeError(
                "LLM not configured. Pass `llm` option to SpiderBrowserOptions for AI methods."
            )
